polygon/blocksync/block_downloader.go

```diff
@@ -0,0 +1,132 @@
+package blocksync
+
+import (
+	"context"
+	"errors"
+	"math/big"
+
+	"github.com/ledgerwatch/erigon/core/types"
+)
+
+func NewBlockDownloader(p2pLayer p2pLayer, dbLayer dbLayer, heimdallLayer heimdallLayer) *BlockDownloader {
+	return &BlockDownloader{
+		p2pLayer:      p2pLayer,
+		dbLayer:       dbLayer,
+		heimdallLayer: heimdallLayer,
+	}
+}
+
+type BlockDownloader struct {
+	p2pLayer      p2pLayer
+	dbLayer       dbLayer
+	heimdallLayer heimdallLayer
+}
+
+func (bd BlockDownloader) ForwardDownloadUsingCheckpoints(ctx context.Context, fromBlockNum uint64) error {
+	return bd.downloadUsingCheckpoints(ctx, fromBlockNum, false)
+}
+
+func (bd BlockDownloader) ForwardDownloadUsingMilestones(ctx context.Context, fromBlockNum uint64) error {
+	return bd.downloadUsingMilestones(ctx, fromBlockNum, false)
+}
+
+func (bd BlockDownloader) BackwardDownloadUsingCheckpoints(ctx context.Context, fromBlockNum uint64) error {
+	return bd.downloadUsingCheckpoints(ctx, fromBlockNum, true)
+}
+
+func (bd BlockDownloader) BackwardDownloadUsingMilestones(ctx context.Context, fromBlockNum uint64) error {
+	return bd.downloadUsingMilestones(ctx, fromBlockNum, true)
+}
+
+func (bd BlockDownloader) downloadUsingMilestones(ctx context.Context, fromBlockNum uint64, reverse bool) error {
+	milestones, err := bd.heimdallLayer.FetchMilestones(fromBlockNum)
+	if err != nil {
+		return err
+	}
+
+	err = bd.downloadUsingStatePoints(ctx, statePointListFromMilestones(milestones), reverse)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (bd BlockDownloader) downloadUsingCheckpoints(ctx context.Context, fromBlockNum uint64, reverse bool) error {
+	checkpoints, err := bd.heimdallLayer.FetchCheckpoints(fromBlockNum)
+	if err != nil {
+		return err
+	}
+
+	err = bd.downloadUsingStatePoints(ctx, statePointListFromCheckpoints(checkpoints), reverse)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (bd BlockDownloader) downloadUsingStatePoints(ctx context.Context, spl statePointList, reverse bool) error {
```

Comment:
```suggestion
func (bd BlockDownloader) downloadUsingStatePoints(ctx context.Context, statePoints statePointList, reverse bool) error {
```

let's try to avoid abbreviations and shortenings (unless it is a general practice like `i` in the loop)

Reply:
addressed